File: Assignment_logreg_bank.R
# Logistic Regression Assignment on bank dataset

# Business Problem: To classify if customer will opt for a term deposit or not

# Data collection: 
bank_raw <- read.csv("D:\\bank-full.csv", sep = ";")
# 45211 records of 17 variables
names(bank_raw)
'''  [1] "age"       "job"       "marital"   "education" "default"   "balance"  
 [7] "housing"   "loan"      "contact"   "day"       "month"     "duration" 
[13] "campaign"  "pdays"     "previous"  "poutcome"  "y" '''

head(bank_raw)
str(bank_raw)
attach(bank_raw)
table(job) #12 levesls
table(marital) #3 levesls
table(education) #4 levesls
table(default  ) #2 levesls
table(housing  ) #2 levesls
table(loan     ) #2 levesls
table(contact  ) #3 levesls
table(month    ) #12 levesls
table(poutcome     ) #4 levesls

# dependent variable
table(y     ) #2 levesls
prop.table(table(y))
# no-39922  88.3% ,    yes- 5289  11.70%
# our classes here are imbalanced. 

# to know how R dummifies categorical variable
contrasts(bank_raw$y)
# no is 0 and yes is 1 (i.e., not taking TD is taken as 0)

# checking for missing values
sum(is.na(bank_raw))
library(Amelia)
missmap(bank_raw,main='Observed vs missing values')
# the whole plot is blue (for observed values), no missing values

######splitting the dataset
library(caTools)
set.seed(123)
split_data = sample.split(bank_raw$y,SplitRatio=0.6)
train = subset(bank_raw, split_data==T)
test <- subset(bank_raw, split_data==FALSE)

# lets check the proportions of yes and no of y variable
prop.table(table(train$y)) # no is 0.883 and yes is 0.1169
prop.table(table(test$y))  # no is 0.8829 and yes is 0.117
prop.table(table(bank_raw$y))   # no is 0.883 and yes is 0.1169
# the proportions of both classes is same in train, test and original datasets

# we have seen that the data is imbalanced, lets build a model using above 
# random sampling and see the accuracy of this imbalanced dataset

logreg1 <- glm(y ~., data=train,family='binomial')
pred1 <- predict(logreg1, newdata=test)
pred1[1:5]

library(ROSE)
accuracy.meas(test$y, pred1)
'''
precision: 0.689 means the model has correcctly predicted 70%  of its positive 
predictions
recall: 0.263 means out of acctual positives , mmodel has only predicted 26% as
positives
F: 0.190 this balances precision and recall. But only 19%, so very less
'''
confusion1 <- table(test$y, pred1>=0.5)
table(test$y)
accuracy1 <- sum(diag(confusion1))/sum(confusion1)
accuracy1
# though accuracy is 90%, the results are highly deceiving since minority class 
# holds minimum effect on overall accuracy.

################### Handling imbalanced data
''' ML algorithms believe that the datasset has balanced class distributions.
When we have imbalanced datasets, the results get biased towars majority class.
There are various techniques to deal with such datasets. we will use them and
build our mmodels and evaluate their performances .
We will be using ROSE package. We will use train dataset to build the newdata'''

library(ROSE)
head(train)
# OVERSAMPLING
# we oversample the minority class
table(train$y)
# no=23953, yes=3173, total=27126
23953+3173
23953+23953 # 47906
bank_over <- ovun.sample(y~.,data=train,method='over',N=47906,seed=123)$data
table(bank_over$y)
#    no   yes 
#  23953 23953

# Undersampling: we undersample majority class without replacement
3173+3173 #6346
bank_under <- ovun.sample(y~.,data=train,method='under',N=6346,seed=123)$data
table(bank_under$y)

# both
bank_both <- ovun.sample(y~.,data=train,method='both',N=nrow(train),
                         p=0.5,seed=123)$data
table(bank_both$y)
#    no   yes 
#  13713 13413  total=27126 same as total records in train data

# Synthetic data
bank_rose <- ROSE(y~., data=train, seed=123)$data
table(bank_rose$y)
#    no   yes 
#  13713 13413

#### model building using the new datasets
# computing model using each data and evaluating its accuracy

logreg_over <- glm(y ~., data=bank_over,family='binomial')
logreg_under <- glm(y ~., data=bank_under,family='binomial')
logreg_both <- glm(y ~., data=bank_both,family='binomial')
logreg_rose <- glm(y ~., data=bank_rose,family='binomial')

# making predictions using test data
pred_over <- predict(logreg_over, newdata=test)
pred_under <- predict(logreg_under, newdata=test)
pred_both <- predict(logreg_both, newdata=test)
pred_rose <- predict(logreg_rose, newdata=test)

# ROC and AUC
# plotting all ROC curves
roc.curve(test$y, pred_over) # AUC is 0.910
roc.curve(test$y, pred_under,add=TRUE,col='green') # AUC is 0.908
roc.curve(test$y, pred_both,add=TRUE,col='red') # AUC is 0.908
roc.curve(test$y, pred_rose,add=TRUE,col='blue') # AUC is 0.904
legend(0.6,0.45, c('over','under','both','rose'),lty=c(1,1),
       lwd=c(2,2),col=c('black','green','red','blue'))

# we are getting almost same accuracy from all sampling methods. Still
# highest accuracy is from oversampling technique

bank_holdout <- ROSE.eval(y ~ ., data =train,learner = glm,method.assess="holdout",
                          control.learner=list(family=binomial),seed = 123)
bank_holdout # auc is 0.908
''' We see that our accuracy retains at 0.908 and shows that our predictions
are not suffering from high variance.'''

'''
CONCLUSIONS

The business problem is to find if the customer has subscribed to a term deposit 
or not. Since the output variable is categorical we have used logistic 
regression technique.

We have found that the dataset is imbalanced. So we have used various sampling
methods and We have prepared various models. The accuracy was almost same
with all sampling techniques. But highest accuracy was obtained from data 
prepared using oversampling technique.  

We used holdout method to check if predictions have high variance. We obtained
same accuracy, suggesting that our model is good.

'''

File: Assignment_logreg_bank.py
#### LOGISTIC REGRESSION ASSIGNMENT - BANK dataset

# Business Problem : To classify if customer will subscribe or not to a term deposit

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix,accuracy_score,precision_score,recall_score,f1_score

# EDA
bank_raw = pd.read_csv("D:\\bank-full.csv",header=0,sep=';')
bank_raw.columns
''' 'age', 'job', 'marital', 'education', 'default', 'balance', 'housing',
       'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'y' 
numeric var: age, balance, day,duration,campaign, pdays, previous
categorical var: 'job', 'marital', 'education', 'default', 'housing',
       'loan', 'contact', 'month','poutcome', 
dependent variable: 'y' - whether taken TD (Term Deposit) or not      
       '''
bank_raw.head()
bank_raw.describe()
bank_raw.dtypes
 
bank_raw.y.value_counts()
''' no     39922
   yes     5289 '''

sns.countplot(x='y', data=bank_raw)

y_no = len(bank_raw[bank_raw['y']=='no'])
y_yes = len(bank_raw[bank_raw['y']=='yes'])
y_no/(y_no+y_yes) #  88.3%
y_yes/(y_no+y_yes)    # 11.69%
# one class has very low proportion, this is an imbalanced dataset

bank_raw.groupby('y').mean()
''' the customers having more age have taken TD.
Customers who have maintained more balance in their accounts have taken TD
Customers who were contacted lesser number of times in this campaign have taken TD.'''

bank_raw.groupby('job').mean()
pd.crosstab(bank_raw.job, bank_raw.y).plot(kind='bar')
''' based on job type, we can know if TD is taken or not. For example, persons 
in management cadre, blue collar jobs, technical jobs, admin and services should be
given more importance in the campaigns.'''

bank_raw.age.hist();plt.xlabel('Age');plt.ylabel('Frequency');
# most of the bank customers age is between 26 to 40 years (for this dataset)


# To get the count of null values in the data 

bank_raw.isnull().sum() # no missing values

########## creating dummy columns for the categorical columns 
#  categorical var: 'job', 'marital', 'education', 'default', 'housing',
#  'loan', 'contact', 'month','poutcome'

bank_dummies = pd.get_dummies(bank_raw[['job', 'marital', 'education', 'default', 'housing','loan', 'contact', 'month','poutcome']])
# 44 dummy varaibles have been created out of 9 categorical variables

# Dropping the columns for which we have created dummies
# copying bank_raw in a new dataset 
bank_raw1 = bank_raw.copy()
bank_raw1.drop(['job', 'marital', 'education', 'default', 'housing','loan', 'contact', 'month','poutcome'], inplace=True, axis=1)

# adding the dummy variables to the bank_raw1 data frame 
bank = pd.concat([bank_raw1, bank_dummies],axis=1)

bank.columns
# 7th column is 'y'
bank.iloc[:6,7]

# splitting dataset
from sklearn.model_selection import train_test_split
train, test = train_test_split(bank,test_size=0.3, random_state=123)

train_X = train.loc[:,train.columns != 'y']
train_y = train.loc[:,train.columns == 'y']
test_X = test.loc[:, test.columns != 'y']
test_y = test.loc[:, test.columns == 'y']

train_y.y.value_counts()
3723/31647
test_y.y.value_counts()
1566/13564
# percentage of y is 12% in train and test dataset. 
# If we were to predict all as 'no', even then we will get 82% accuracy.

''' This is an imbalanced dataset. The results of logistic regression will get biased 
towards majority class (no TD). So before model building we have to balance the dataset. '''

# lets build model on this imbalanced dataset.
# train the model
logreg_imbalanced = LogisticRegression().fit(train_X, train_y)
pred = logreg_imbalanced.predict(test_X)
pred = pd.DataFrame(pred)

confusion_matrix(pred,test_y)
#      [[11706,  1059],
#      [  292,   507]]
11706/13564
# Here only if we consider correctly predicted TN (11706), accuracy is 86%
accuracy_score(pred,test_y) # 90% 
precision_score(pred,test_y,pos_label='yes') # 0.3238 out of predicted positives, 32% are correctly predicted
recall_score(pred,test_y,pos_label='yes') # 0.6345 out of actual postives, 63% are correctly predicted by the model
f1_score(pred,test_y,pos_label='yes') # 0.4288 this balances precision and recall

######################## OVERSAMPLING of minority class
# we have already split data into train and test.  let us use train data for oversampling

# now separate minority and majority classes
no_TD = train[train.y == 'no']
yes_TD = train[train.y =='yes']
 
# oversampling the minority class
from sklearn.utils import resample

bank_over = resample(yes_TD, replace=True, n_samples=len(no_TD),random_state=123)

# combining majority and oversampled minority
oversampled = pd.concat([no_TD, bank_over])

# checking new class counts
oversampled.y.value_counts()
# yes    27924
# no     27924

######################## Model building using oversampled data
X_train_over = oversampled.drop('y',axis=1)
y_train_over = oversampled.y

logreg_oversampled = LogisticRegression().fit(X_train_over,y_train_over)

pred_over = logreg_oversampled.predict(test_X)

# accuracy measurements
confusion_matrix(pred_over,test_y)
# 10189,   296
# 1809,  1270
accuracy_score(pred_over,test_y) # 0.8448 previously 90%
precision_score(pred_over,test_y,pos_label='yes') # 0.8109 out of predicted positives, 81% are correctly predicted
# Now it is 81% (previous model just 32%)
recall_score(pred_over,test_y,pos_label='yes') # 0.4125 out of actual postives, 41% are correctly predicted by the model
# Now it is 41% (previous model 63%)
f1_score(pred_over,test_y,pos_label='yes') # 0.5468 
# Now it is 55% (previous model 43%)

''' Though accuracy has decreased, f1_score has increased. Also the models prediction ability
has increased.'''

###################### Generating Synthetic samples using SMOTE

from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import LabelEncoder

sm = SMOTE(random_state=123, ratio=1.0)
X_train_smote, y_train_smote = sm.fit_sample(train_X,train_y)
# 55848 obs in each 

classifier = LogisticRegression()
logreg_smote = LogisticRegression().fit(X_train_smote,y_train_smote)

pred_smote = logreg_smote.predict(test_X)

# accuracy measurements
confusion_matrix(pred_smote,test_y)
# 10713, 614
# 1285,  952
accuracy_score(pred_smote,test_y) # 0.86 
precision_score(pred_smote,test_y,pos_label='yes') # 0.6079 
recall_score(pred_smote,test_y,pos_label='yes') # 0.4256 
f1_score(pred_smote,test_y,pos_label='yes') # 0.5007 


'''
            logreg_imbalanced    logreg_oversampled   logreg_smote
accuracy          0.90                 0.84               0.86
precision         0.32                 0.81               0.61 
recall            0.63                 0.41               0.43  
f1                0.43                 0.55               0.50

Accuracy and recall scores are better in model with imbalanced data
Precision and f1 scores are better in oversampled model
Overall, oversampled model is better.

CONCLUSIONS

We have to classify customers those who have subscribed to Term deposits from 
those who have not. Since the output variable is categorical (yes, no), we have 
used Logistic Regression technique.

From exploratory data analysis, we have found that the dataset is highly imbalanced.
So we have tried sampling techniques: oversamping and synthetic sampling. 
We have split the data into train and test: on train data we have done sampling.

We have run 3 different models: one with imbalanced data and others using oversampling
and SMOTE. The model built using oversampling technique is better. 

Due to time limitation we have tried only 3 models. We can build models using 
undersampling, tree-based algorithms, Cost-based algorithm, change the split ratio 
for train-test, change random_state, etc. Then choose the best model.
 
'''



















